apps/projects/routes.py

```diff
@@ -924,146 +989,153 @@ def post(self, project_id):
+        )
+        mimetype = app.config.get('CODEC_MIMETYPE_MAP')[metadata.get('codec_name')]
         storage_id = app.fs.put(
-            stream, thumbnail_filename, None,
-            asset_type='thumbnails', storage_id=doc['storage_id'], content_type='image/png'
+            content=file_stream,
+            filename=thumbnail_filename,
+            project_id=None,
+            asset_type='thumbnails',
+            storage_id=self._project['storage_id'],
+            content_type=mimetype
         )
-        doc = app.mongo.db.projects.find_one_and_update(
-            {'_id': doc['_id']},
+
+        # delete old file
+        if self._project['thumbnails']['preview'] \
+                and storage_id != self._project['thumbnails']['preview']['storage_id']:
+            app.fs.delete(self._project['thumbnails']['preview']['storage_id'])
+
+        # save new thumbnail info
+        self._project = app.mongo.db.projects.find_one_and_update(
+            {'_id': self._project['_id']},
             {'$set': {
-                'preview_thumbnail': {
+                'thumbnails.preview': {
                     'filename': thumbnail_filename,
-                    'url': get_url_for_media(doc.get('_id'), 'thumbnail'),
                     'storage_id': storage_id,
-                    'mimetype': 'image/png',
+                    'mimetype': mimetype,
                     'width': metadata.get('width'),
                     'height': metadata.get('height'),
                     'size': metadata.get('size'),
+                    'position': 'custom'
                 }
             }},
             return_document=ReturnDocument.AFTER
         )
-        return json_response(doc)
+        add_urls(self._project)
+
+        return json_response(self._project['thumbnails']['preview'])
+
+    def _get_timeline_thumbnails(self, amount):
+        """
+        Get list or create thumbnails for timeline
+        :param amount: amount of thumbnails
+        :type amount: int
+        :return: json response
+        :rtype: flask.wrappers.Response
+        """
+        # resource is busy
+        if self._project['processing']['thumbnails_timeline']:
+            return json_response({"processing": True}, status=202)
+        # no need to generate thumbnails
+        elif amount == len(self._project['thumbnails']['timeline']):
+            return json_response(self._project['thumbnails']['timeline'])
+        else:
+            # set processing flag
+            self._project = app.mongo.db.projects.find_one_and_update(
+                {'_id': self._project['_id']},
+                {'$set': {'processing.thumbnails_timeline': True}},
+                return_document=ReturnDocument.AFTER
+            )
+            # run task
+            generate_timeline_thumbnails.delay(
+                json_util.dumps(self._project),
+                amount
+            )
+            return json_response({"processing": True}, status=200)
+
```

Comment:
We return 202 in next cases:
- when you send a edit video request, but video is in processing at the moment https://github.com/superdesk/video-server/blob/073d6bea263b45d29f2190535322a7fe684c7d14/apps/projects/routes.py#L581
- when you send a duplicate video request, but video or thumbnails are in processing at the moment  https://github.com/superdesk/video-server/blob/073d6bea263b45d29f2190535322a7fe684c7d14/apps/projects/routes.py#L789
- when you send a custom preview thumbnail file upload request, but thumbnail preview is in processing at the moment https://github.com/superdesk/video-server/blob/073d6bea263b45d29f2190535322a7fe684c7d14/apps/projects/routes.py#L1030
- when you send a get/capture timeline/preview thumbnails request, but timeline/preview generation is still in processing https://github.com/superdesk/video-server/blob/073d6bea263b45d29f2190535322a7fe684c7d14/apps/projects/routes.py#L1082
https://github.com/superdesk/video-server/blob/073d6bea263b45d29f2190535322a7fe684c7d14/apps/projects/routes.py#L1110
- when you try to get a video stream, but video is in processing https://github.com/superdesk/video-server/blob/073d6bea263b45d29f2190535322a7fe684c7d14/apps/projects/routes.py#L1166

IMO in other cases there is not sense to return 202, for example if video is editing you still can fetch list, video details, thumbnail and etc. By fetching video details you can actually see a processing status.
BTW. Does it makes sense to create an endpoint to fetch video status? For now you can do it only by fetching entire project details.

Reply:
I see the definition of `http code 202` is the request has been accepted for processing (https://httpstatuses.com/202). That mean If it return `202` in these cases above, maybe It make the client will misunderstand that the task was accepted and put into queue, but in reality, it was rejected.
So IMO, all cases we edit or get thumbnail when the processing of task is still running. It should return a 4xx code (maybe 406 Not acceptable or 409 Conflict) to make sure the client know that the request has not fulfilled (the task was rejected, not add to queue).
And all cases when task was accepted and was put to queue, but it's not finished yet, we should return 202 instead of 200. 
eg: the case we edit video when `processing` is `false`
https://github.com/superdesk/video-server/blob/073d6bea263b45d29f2190535322a7fe684c7d14/apps/projects/routes.py#L658
How do you think about this opinion?

Reply:
About the new endpoint for fetch video status. It's okay, when we get status by fetching entire project details.

Reply:
Sounds good to me, I will update it.